File: doc/ex-20200806-hello-world-rules/potato.js
"use strict";
function nextAvailableDates(date, asOf, limit) {
    return legalDates(asOf)
        .filter(x => x >= date)
        .splice(0, limit);
}
function legalDates(asOf) {
    return fullMoonDates(asOf)
        .flatMap(x => [x, addDays(x, 1)]);
}
function fullMoonDates(asOf) {
    return getDatesFromURL("https://www.almanac.com/astronomy/moon/full/");
}
function getDatesFromURL(_) {
    return (["2020-01-10",
        "2020-02-08",
        "2020-03-09",
        "2020-04-07",
        "2020-05-06",
        "2020-06-05",
        "2020-07-04",
        "2020-08-02",
        "2020-09-01",
        "2020-10-01",
        "2020-10-30",
        "2020-11-29",
        "2020-12-29"].map(x => new Date(x)));
}
console.log("the next available dates are: " + JSON.stringify(nextAvailableDates(new Date("2020-08-06"), new Date("2020-08-06"), 4)));
var Deontic;
(function (Deontic) {
    Deontic["must"] = "MUST";
    Deontic["may"] = "MAY";
    Deontic["shant"] = "SHANT";
})(Deontic || (Deontic = {}));
;
function rule1(world, party, action, actionSpec, exemption) {
    if (action === "trade" && isPotato(actionSpec)) {
        if (isLegalPotato(actionSpec, world.legalDates) ||
            (exemption != undefined && exemption.from == "DirectorOfAgriculture")) {
            return Deontic.may;
        }
        return Deontic.shant;
    }
}
function isPotato(item) {
    return item.hasOwnProperty("species") && item.species === "Tuberosum solanum";
}
function isLegalPotato(item, legalDates) {
    return item.hasOwnProperty("wasPlanted") && legalDates.includes(item.wasPlanted);
}
const potato1 = { species: "Solanum tuberosum", wasPlanted: new Date("2020-01-10") };
const potato2 = { species: "Solanum tuberosum", wasPlanted: new Date("2020-01-11") };
const potato3 = { species: "Solanum tuberosum", wasPlanted: new Date("2020-01-12") };
console.log(`potato1: ` + rule1({ legalDates: legalDates(new Date("2020-08-06")) }, "_", "trade", potato1, undefined));
console.log(`potato2: ` + rule1({ legalDates: legalDates(new Date("2020-08-06")) }, "_", "trade", potato2, undefined));
function addDays(date, days) {
    var result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
}

File: doc/ex-20200806-hello-world-rules/potato.pl
:- use_module(library(clpfd)).
:- use_module(library(julian)).
:- use_module(library(yall)).
rule(1, P, may, trade(Item))   :- rule(_,isPotato(Item)), rule(_, isLegalPotato(Item)).
rule(1, P, may, trade(Item))   :- rule(_,isPotato(Item)), hasExemption(P, from(directorOfAgriculture), that(P,may,trade(Item))).
rule(1, P, shant, trade(Item)) :- rule(_,isPotato(Item)), \+ rule(1, P, may, trade(Item)).
rule(2, isPotato(Item)) :- member(Item.get(species), ["Solanum tuberosum"]).
rule(3, isLegalPotato(Item)) :- rule(_,isPotato(Item)), rule(_,legalDates(Item.get(wasPlanted))).
rule(4,
     legalDates(Date)) :- rule(_, fullMoonDates(FMD)),
                          ( Date = FMD ; 
                            form_time(FMD,datetime(J0,_)), J1 #= J0+1, form_time(Date,datetime(J1,0)) ).
rule(5, fullMoonDates(2020-1-10)).
rule(5, fullMoonDates(2020-2-8)).
rule(5, fullMoonDates(2020-3-9)).
rule(5, fullMoonDates(2020-4-7)).
rule(5, fullMoonDates(2020-5-6)).
rule(5, fullMoonDates(2020-6-5)).
rule(5, fullMoonDates(2020-7-4)).
rule(5, fullMoonDates(2020-8-2)).
rule(5, fullMoonDates(2020-9-1)).
rule(5, fullMoonDates(2020-10-1)).
rule(5, fullMoonDates(2020-10-30)).
rule(5, fullMoonDates(2020-11-29)).
rule(5, fullMoonDates(2020-12-29)).
potato1(item{species:"Solanum tuberosum", wasPlanted:2020-1-10}).
potato2(item{species:"Solanum tuberosum", wasPlanted:2020-1-11}).
potato3(item{species:"Solanum tuberosum", wasPlanted:2020-1-12}).
hasExemption(nobody, from(no-one), that(nothing)).
